### PR TITLE
QTable: implementation of Quasar's Qtable element

### PR DESCRIPTION
**Pull request by Diegiwg**
After implementing some elements (my other PRs), I managed to understand the minimum to 'correctly' implement this element.

At the moment, the 'filter/search' API doesn't seem like the best choice, but I don't know how to make it better.

About `qtable.Column`, I didn't know how to define which fields are needed and which fields can be ignored in such a complex object, so I decided to make a wrapper class, which would handle everything.

About the use of `Eval` in Fields: I don't know how dangerous it can be to use eval, however, this was the only method I could think of to transform JS functions 'from string to real objects'.

#413 #370

### CODE REVIEW

**Review by Diegiwg — COMMENTED** [review]
Now that I realized, that I forgot to write the DocString.

But I'm exhausted, if the implementation is 'good', let's do it.

File: nicegui/elements/qtable.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+class QTable(Element):
+    def __init__(
+        self,
+        columns: Optional[dict] = [],
+        rows: Optional[list] = [],
+        title: Optional[str] = None,
+        selection_mode: Optional[Literal['single', 'multiple', 'none']] = 'none',
+        selection_key: Optional[str] = '',
+        filter_config: Optional[dict] = {
```
Comment (dclause) [inline nicegui/elements/qtable.py:19]: May I suggest to not use mutable data structure as default argument ? 
_see flake8-bugfear rule "B006 Do not use mutable data structures for argument defaults "
here for details: https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/_

Plus, see the 'v-if' comment : allow None by default would allow to remove the slot which is not possible here.

File: nicegui/elements/qtable.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            :row-key="this.selection_key"
+            :selection="this.selection_mode"
+
+            v-model:selected="selected"
+            @selection="onSelection"
+
+            :filter="filter"
+        >
+
+        <template v-slot:top-right>
```
Comment (dclause) [inline nicegui/elements/qtable.js:32]: I would suggest to add a v-if on that slot here : unless I am wrong, it means here that the slot is "mandatory" as I can't go without the search bar here.

Comment (dclause) [inline nicegui/elements/qtable.js:32]: Could I suggested to always use the v-slot body syntax ? 
https://quasar.dev/vue-components/table#body-slots

That would be more complex in that PR but allow much more flexibilty. Without giving too much thoughts, maybe something like : 
```
with ui.table():
  with ui.tablerow():
     with ui.tablecolumn('name'):
        ui.label('Here we can have custom elements')
     with ui.tablecolumn('actions'):
        # Here maybe a dropdown menu with 'edit' / 'delete' icons on the row for instance
```

File: nicegui/elements/qtable.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
+function eval_fields(cols) {
+    // For each Field, Sort and Format field within the column definition,
+    // look for '=>' and if so, try to eval to run the function.
+
+    cols.forEach((col) => {
+        if (!col.field || !col.field.includes("=>")) return;
```
Comment (dclause) [inline nicegui/elements/qtable.js:6]: I'm not really sure of the implication here, but I fear that if this acts on user input data, it could be used for some sort of injection. 
I have not been able to find how though. If I change a name "Bob" to something like : 
`{'name': "<script>alert('test');</script>", 'age': {'deep': 21}},`
everything crashes. That is not related to that eval though, so I'm am not sure why at the moment.

File: nicegui/elements/qtable.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        self._props['filter_config'] = filter_config
+
+        self.selected = []
+        self.on('selected', handler=self.handle_selected_event)
+
+    def set_data(self, data):
+        self._props['rows'] = data
+
+        self.update()
+        self.run_method('evalFunction')
```
Comment (dclause) [inline nicegui/elements/qtable.py:64]: I get it regarding the context, but this fills hacky.

I am wondering if this is needed at all. I mean, we are injecting the data from python code. Therefore we can prepare them through a transformer method that would (in your example here) unpack the age.deep to age and make the `row => row.age.deep useless.`
I don't see anything that this feature allow that we could not do without.

File: website/reference.py [hunk 5]
```diff
@@ -276,6 +276,32 @@ def update():
         ui.button('Update', on_click=update)
         ui.button('Select all', on_click=lambda: table.call_api_method('selectAll'))
 
+    @example(ui.qtable, menu)
+    def qtable_example():
+        fields = [
+            {'name': 'name', 'field': 'name', 'label': 'Name', 'required': True},
+            {'name': 'age', 'label': 'Age', 'field': 'row => row.age.deep', 'sortable': True},
```
Comment (dclause) [inline website/reference.py:283]: Let me suggest that this should be done in python. 
Instead of using a JS method client side, prepare your data accordingly server-side in python.

File: website/reference.py [hunk 6]
```diff
@@ -276,6 +276,32 @@ def update():
...
+            {'name': 'Carol', 'age': {'deep': 42}},
+        ]
+
+        table = ui.qtable(
+            columns=fields,
+            rows=data,
+            selection_mode='single',
+            selection_key='name',
+            title='QTable',
+            filter_config={'label': 'Search...', 'icon': 'search', 'color': 'primary', 'dense': True})
```
Comment (dclause) [inline website/reference.py:298]: See my v-if suggestion above and the B0002 comment to have it to None by default to remove the search bar at all.

Comment (dclause) [inline website/reference.py:298]: Same could applies for the bottom slot that we could also allow customization and 'None' to disable.

File: nicegui/elements/qtable.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    },
+
+    methods: {
+        onSelection() {
+            setTimeout(() => {
+                this.$emit("selected", this.selected);
+            }, 0);
+        },
+
+        evalFunction() {
```
Comment (dclause) [inline nicegui/elements/qtable.js:71]: Let's remove this and use python to prepare the data rather then JS transformation here.
Make sense for Quasar API to consume API data, but we have a whole python layer to make our transformation here.

**Review by dclause — CHANGES_REQUESTED** [review]
This is my first github review actually ! Hope I made it right :)

**Comment by falkoschindler** [comment]
I finally found the time to look into this PR. Thanks a lot for your plenty contributions, @Diegiwg!

My main concern is: Why do we need a custom JS component at all? We can assemble and remotely control Quasar elements from Python like we do with other NiceGUI elements. Introducing a custom component results in more code with less features/flexibility. Filtering might be a reason for client-side code. But setting the `filter` prop to a string should work just fine for most cases.

I think @dclause's PR #500 is a great successor. But let's discuss if there are arguments for this PR #434.

**Comment by falkoschindler** [comment]
We ended up working on #500 to make it production ready. Therefore I'll close this PR.
